Kuadrant/kuadrant-operator
# Reconcile Sub Component Limitador CR

Part of: https://github.com/Kuadrant/kuadrant-operator/issues/307
Closes: https://github.com/Kuadrant/kuadrant-operator/issues/163
RFC: https://github.com/Kuadrant/architecture/blob/main/rfcs/0006-kuadrant_sub_components_configurations.md

The changes here allow a user to configure the setting in the Limitador CR from the kuadrant CR. 

# Verification
* Deploy this branch using `make local-setup`.
* Wait for deploy to complete and all resources to be a ready state.
* Deploy the custom manifest.
  * `kubectl apply -f https://raw.githubusercontent.com/Boomatang/k8s_configs/kuadrant-operator-PR350/manifest/kuadrant_spec_limitador.yaml
`
  * The custom manifest will deploy an instance of redis along with all the resources to test the different configurations.
  * The Kuadrant CR has every [spec](https://github.com/Boomatang/k8s_configs/blob/kuadrant-operator-PR350/manifest/kuadrant_spec_limitador.yaml#L129-L158) that this PR allows configuring set.
  * The manifest is linked to a tag [kuadrant-operator-PR350](https://github.com/Boomatang/k8s_configs/tree/kuadrant-operator-PR350) to ensure state does not change.
* Wait of deployment to complete
* The spec that is being reconciled are listed in the RFC: [Limitador Spec](https://github.com/Kuadrant/architecture/blob/main/rfcs/0006-kuadrant_sub_components_configurations.md#limitador-spec)
* Editing the spec directly  in the Limitador CR should be reverted to the spec in the Kuadrant CR
* Editing the spec in the Kuadrant CR should update the spec in the Limitador CR
  * **NOTE:** fully removing a spec from the kuadrant CR will not remove it from the Limitador CR. This is do to allow upgrades. A user may have custom spec already in the Limitador CR. This change only overrides the Limitador CR spec if that spec is defined in Kuadrant CR.

## Use Case:
### Existing Limitador Users
It is possible that a user can be using limitador as a standalone product before using kuadrant. In the case where the user named the limitador CR `limitador` and places the kuadrant CR in the same namespace, the kuadrant operator will take ownership of the limitador CR.  This will be done by setting the ownerReference in the limitador CR.

Validating this use case.
* Follow the verification steps to get the environment set up.
* Scale down the kuadrant operator
* Edit the limitador CR 
  * Set spec as follows
  ```yaml
  spec:
    replicas: 1
  ```
  * Remove the `ownerReference` block from the metadata.
* Scale back up the kuadrant operator
* Expected:
  * The `spec` switches to that defined in the kuadrant CR
  * The `ownerReference` is set to the kuadrant operator
